### PR TITLE
Gemma 2

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do ?

Add Gemma 2 support 

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -444,6 +446,12 @@ def model_provider_func(self, pre_process, post_process):
             )
             if self.cfg.get("apply_embedding_scaling", False) and parallel_state.is_pipeline_first_stage():
```
Comment (yaoyu-33) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:447]: can we move these extend instance (for gemma1 and 2) to a method maybe called "def model_customize(self)"? and here only do self.model_customize. Please use your ideas if you have better naming.

File: nemo/collections/nlp/models/language_modeling/megatron/gemma2/gemma2_modules.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        # [b, np, sq, hn] --> [sq, b, np, hn]
+        context = context.permute(2, 0, 1, 3).contiguous()
+
+        # [sq, b, np, hn] --> [sq, b, hp]
+        new_context_shape = context.size()[:-2] + (self.hidden_size_per_partition,)
+        context = context.view(*new_context_shape)
+        return context
+
+
+class TERowParallelLinearLayerNorm(TERowParallelLinear):
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/nlp/models/language_modeling/megatron/gemma2/gemma2_modules.py:244]: Gemma 2 adds post-attention and post-mlp layernorms to the model. This is one way to implement it using mcore specs without modifying mcore code

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 1 above)
Comment (cuichenx) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:447]: sure that's a good idea

**Review by yaoyu-33 — APPROVED** [review]

**Comment by cuichenx** [comment]
closing this in favor of #9672

**Pull request closed without merging**